### PR TITLE
Change consumer flow

### PR DESCRIPTION
**Pull request by DavidFair**

### CODE REVIEW

File: OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py [hunk 1]
```diff
@@ -82,54 +76,57 @@ def setup_requests(url, method, desc):
     return response.text
 
 
-def aq_make(hostname: str, aq_fields: AqFields):
-    logger.debug("Attempting to make templates for %s", hostname)
-
+def aq_make(addresses: List[OpenstackAddress], os_data: OsDescription):
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py:79]: ^^

File: OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py [hunk 2]
```diff
@@ -82,54 +76,57 @@ def setup_requests(url, method, desc):
...
 
-def aq_manage(hostname, env_type, env_name):
-    logger.debug("Attempting to manage %s to %s %s", hostname, env_type, env_name)
+        url = ConsumerConfig().aq_url + MAKE_SUFFIX.format(hostname)
+        setup_requests(url, "post", "Make Template: ", params)
 
-    url = ConsumerConfig().aq_url + MANAGE_SUFFIX.format(hostname, env_type, env_name)
 
-    setup_requests(url, "post", "Manage Host")
+def aq_manage(addresses: List[OpenstackAddress]):
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py:103]: ^^

File: OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py [hunk 3]
```diff
@@ -82,54 +76,57 @@ def setup_requests(url, method, desc):
...
+        setup_requests(url, "post", "Manage Host", params=params)
 
-def create_machine(message, hostname, prefix):
-    logger.debug("Attempting to create machine for %s ", hostname)
 
-    vcpus = message.get("payload").get("vcpus")
-    memory_mb = message.get("payload").get("memory_mb")
-    uuid = message.get("payload").get("instance_id")
-    vmhost = message.get("payload").get("host")
+def create_machine(message: RabbitMessage, vm_data: VmData) -> str:
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py:117]: We need a description for the method

File: OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py [hunk 4]
```diff
@@ -142,66 +139,61 @@ def delete_machine(machinename):
 
 
 def create_host(
-    hostname,
-    machinename,
-    firstip,
-    osname,
-    osversion,
+    os_details: OsDescription, addresses: List[OpenstackAddress], machine_name: str
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py:142]: Could we add a docstring and a type for the variable returned?

File: OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py [hunk 5]
```diff
@@ -142,66 +139,61 @@ def delete_machine(machinename):
...
 
-    # reset personality etc ...
-    setup_requests(url, "put", "Host Create")
+        logger.debug("Attempting to create host for %s ", address.hostname)
+        url = config.aq_url + f"/host/{address.hostname}"
+        setup_requests(url, "put", "Host Create", params=params)
 
 
-def delete_host(hostname):
+def delete_host(hostname: str):
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py:162]: ^^

File: OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py [hunk 6]
```diff
@@ -142,66 +139,61 @@ def delete_machine(machinename):
...
-    logger.debug(
-        "Attempting to add interface %s to machine %s ", interfacename, machinename
-    )
-
-    url = ConsumerConfig().aq_url + ADD_INTERFACE_SUFFIX.format(
-        machinename, interfacename, macaddr
-    )
-
-    setup_requests(url, "put", "Add Machine Interface")
+def add_machine_nics(machine_name, addresses: List[OpenstackAddress]):
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py:168]: ^^

File: OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py [hunk 7]
```diff
@@ -142,66 +139,61 @@ def delete_machine(machinename):
...
+            "fqdn": address.hostname,
+        }
 
-    setup_requests(url, "put", "Add Machine Interface Address")
+        url = ConsumerConfig().aq_url + "/interface_address"
+        setup_requests(url, "put", "Add Machine Interface Address", params=params)
 
 
-def update_machine_interface(machinename, interfacename):
+def set_interface_bootable(machinename, interfacename):
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/aq_api.py:196]: ^^

File: OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py [hunk 8]
```diff
@@ -1,281 +1,134 @@
...
-    returnstring = message.get("payload").get("metadata").get(key)
-    if returnstring is None:
-        returnstring = message.get("payload").get("image_meta").get(key)
-        if returnstring is None:
-            returnstring = message.get("payload").get("metadata").get(key.lower())
-            if returnstring is None:
-                returnstring = message.get("payload").get("image_meta").get(key.lower())
-    return returnstring
-
+def consume(message: RabbitMessage):
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py:37]: Could we add a description for this method and the type of variable is returned?

File: OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py [hunk 9]
```diff
@@ -1,281 +1,134 @@
...
+        handle_create_machine(message)
 
-    if event == "compute.instance.delete.start":
-        _handle_machine_delete(message)
+    if message.event_type == "compute.instance.delete.start":
+        handle_machine_delete(message)
 
 
-def _handle_create_machine(message):
+def handle_create_machine(rabbit_message: RabbitMessage):
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py:49]: ^^

File: OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py [hunk 10]
```diff
@@ -1,281 +1,134 @@
...
-    )
+    vm_data = VmData.from_message(rabbit_message)
+    os_details = is_aq_managed_image(rabbit_message)
+    network_details = openstack_api.get_server_networks(vm_data)
 
-    if not aq_details.hostnames:
+    if not network_details or not network_details[0].hostname:
+        vm_name = rabbit_message.payload.vm_name
         logger.info("Skipping novalocal only host: %s", vm_name)
         return
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py:60]: Is the method meant to return anything?

File: OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py [hunk 11]
```diff
@@ -1,281 +1,134 @@
...
-        fields.hostnames[0],
-        machinename,
-        firstip,
-        fields.osname,
-        fields.osversion,
-    )  # osname needs to be valid otherwise it fails - also need to pass in sandbox
-    openstack_api.update_metadata(
-        fields.project_id, vm_id, {"AQ_MACHINENAME": machinename}
+
+def _print_debug_logging(rabbit_message: RabbitMessage) -> None:
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py:80]: Could we add a description for this method?

File: OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py [hunk 12]
```diff
@@ -1,281 +1,134 @@
...
-    logger.debug("Creating machine")
-    try:
-        machine_name = aq_api.create_machine(
-            message,
-            hostnames[-1],
-            ConsumerConfig().aq_prefix,
-        )
-    except Exception as err:
-        raise RuntimeError("Failed to create machine") from err
+def handle_machine_delete(rabbit_message: RabbitMessage):
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py:92]: Could we add a description for this method and the type of variable is returned?

File: OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py [hunk 13]
```diff
@@ -1,281 +1,134 @@
...
+        aq_api.delete_host(host.hostname)
+
+    aq_api.delete_machine(rabbit_message.payload.metadata.machine_name)
+
+    logger.info(
+        "=== Finished Aquilon deletion hook for VM %s ===", vm_data.virtual_machine_id
+    )
+
+
+def add_hostname_to_metadata(fields: VmData, network_details: List[OpenstackAddress]):
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py:114]: ^^

File: OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py [hunk 14]
```diff
@@ -1,281 +1,134 @@
...
-    logger.debug("Hostnames: %s", hostnames)
-    return hostnames
+        return
+
+    hostnames = [i.hostname for i in network_details]
+    metadata = {"HOSTNAMES": ",".join(hostnames)}
+    openstack_api.update_metadata(fields, metadata)
 
 
 def on_message(message):
```
Comment (meoflynn) [inline OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py:128]: ^^

**Review by meoflynn — CHANGES_REQUESTED** [review]
Looks good so far, could you add docstrings to the methods and unit tests and also include the type of the value returned by the methods and unit tests?

File: OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py (hunk 10 above)
Comment (DavidFair) [reply OpenStack-Rabbit-Consumer/rabbit_consumer/message_consumer.py:68]: No, we just bail out because we can't handle a machine that's not routable

**Review by meoflynn — CHANGES_REQUESTED** [review]
Looks good to me, just some very small changes we could do for formatting methods

**Review by meoflynn — APPROVED** [review]

**Pull request merged**